transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java

```diff
@@ -181,157 +146,51 @@ public boolean addPollOut(int fd) {
-        while (sqe == 0) {
-            sqe = getSqe();
-
-            if (sqe == 0) {
-                submit();
-                submitted = true;
-            }
-        }
-        setData(sqe, (byte) Native.IORING_OP_POLL_REMOVE, pollMask, fd, 0, 0, 0);
-
-        return submitted;
+        return enqueueSqe(Native.IORING_OP_POLL_REMOVE, 0, fd,
+                convertToUserData(Native.IORING_OP_POLL_ADD, fd, pollMask), 0, 0);
     }
 
     public boolean addConnect(int fd, long socketAddress, long socketAddressLength) {
-        long sqe = 0;
-        boolean submitted = false;
-        while (sqe == 0) {
-            sqe = getSqe();
-
-            if (sqe == 0) {
-                submit();
-                submitted = true;
-            }
-        }
-        setData(sqe, (byte) Native.IORING_OP_CONNECT, 0, fd, socketAddress, 0, socketAddressLength);
-
-        return submitted;
+        return enqueueSqe(Native.IORING_OP_CONNECT, 0, fd, socketAddress, 0, socketAddressLength);
     }
 
     public boolean addWritev(int fd, long iovecArrayAddress, int length) {
-        long sqe = 0;
-        boolean submitted = false;
-        while (sqe == 0) {
-            sqe = getSqe();
-
-            if (sqe == 0) {
-                submit();
-                submitted = true;
-            }
-        }
-        setData(sqe, (byte) Native.IORING_OP_WRITEV, 0, fd, iovecArrayAddress, length, 0);
-
-        return submitted;
-    }
-
-    private int flushSqe() {
-        long kTail = toUnsignedLong(PlatformDependent.getInt(kTailAddress));
-        long kHead = toUnsignedLong(PlatformDependent.getIntVolatile(kHeadAddress));
-        long kRingMask = toUnsignedLong(PlatformDependent.getInt(kRingMaskAddress));
-
-        logger.trace("Ktail: {}", kTail);
-        logger.trace("Ktail: {}", kHead);
-        logger.trace("SqeHead: {}", sqeHead);
-        logger.trace("SqeTail: {}", sqeTail);
-
-        if (sqeHead == sqeTail) {
-            return (int) (kTail - kHead);
-        }
-
-        long toSubmit = sqeTail - sqeHead;
-        while (toSubmit > 0) {
-            long index = kTail & kRingMask;
-
-            PlatformDependent.putInt(arrayAddress + index * INT_SIZE, (int) (sqeHead & kRingMask));
-
-            sqeHead++;
-            kTail++;
-            toSubmit--;
-        }
-
-        //release
-        PlatformDependent.putIntOrdered(kTailAddress, (int) kTail);
-
-        return (int) (kTail - kHead);
+        return enqueueSqe(Native.IORING_OP_WRITEV, 0, fd, iovecArrayAddress, length, 0);
     }
 
     public void submit() {
-        int submitted = flushSqe();
-        logger.trace("Submitted: {}", submitted);
-        if (submitted > 0) {
-            int ret = Native.ioUringEnter(ringFd, submitted, 0, 0);
-            if (ret < 0) {
-                throw new RuntimeException("ioUringEnter syscall");
+        int submit = tail - head;
+        if (submit > 0) {
+            PlatformDependent.putIntOrdered(kTailAddress, tail);
```

Comment:
I think we should add comment as it's not really clear that putIntOrdered is a release memory barrier